blockchain/genesis_test.go

```diff
@@ -26,38 +26,124 @@ import (
+	}{
+		{
+			name:       "cypress chainConfig update",
+			newHFBlock: big.NewInt(3),
+			fn: func(db database.DBManager, newHFBlock *big.Int) (*params.ChainConfig, common.Hash, error) {
+				cypressGenesisBlock.MustCommit(db)
+				cypressGenesisBlock.Config.IstanbulCompatibleBlock = newHFBlock
+				return SetupGenesisBlock(db, cypressGenesisBlock, params.CypressNetworkId, false, false)
+			},
+			wantHash:         params.CypressGenesisHash,
+			wantConfig:       cypressGenesisBlock.Config,
+			wantStoredConfig: cypressGenesisBlock.Config,
+		},
+		{
+			name:       "baobab chainConfig update",
+			newHFBlock: big.NewInt(90909999),
+			fn: func(db database.DBManager, newHFBlock *big.Int) (*params.ChainConfig, common.Hash, error) {
+				baobabGenesisBlock.MustCommit(db)
+				baobabGenesisBlock.Config.IstanbulCompatibleBlock = newHFBlock
+				return SetupGenesisBlock(db, baobabGenesisBlock, params.BaobabNetworkId, false, false)
+			},
+			wantHash:         params.BaobabGenesisHash,
+			wantConfig:       baobabGenesisBlock.Config,
+			wantStoredConfig: baobabGenesisBlock.Config,
+		},
+		{
+			name:       "incompatible config in DB",
+			newHFBlock: big.NewInt(3),
+			fn: func(db database.DBManager, newHFBlock *big.Int) (*params.ChainConfig, common.Hash, error) {
+				// Commit the 'old' genesis block with Istanbul transition at #2.
+				// Advance to block #4, past the Istanbul transition block of customGenesis.
+				genesis := genCypressGenesisBlock()
+				genesisBlock := genesis.MustCommit(db)
+
+				bc, _ := NewBlockChain(db, nil, genesis.Config, gxhash.NewFullFaker(), vm.Config{})
+				defer bc.Stop()
+
+				blocks, _ := GenerateChain(genesis.Config, genesisBlock, gxhash.NewFaker(), db, 4, nil)
+				bc.InsertChain(blocks)
+				// This should return a compatibility error.
+				newConfig := *genesis
+				newConfig.Config.IstanbulCompatibleBlock = newHFBlock
+				return SetupGenesisBlock(db, &newConfig, params.CypressNetworkId, true, false)
+			},
+			wantHash:         params.CypressGenesisHash,
+			wantConfig:       cypressGenesisBlock.Config,
+			wantStoredConfig: params.CypressChainConfig,
+			wantErr: &params.ConfigCompatError{
+				What:         "Istanbul Block",
+				StoredConfig: nil,
+				NewConfig:    big.NewInt(3),
+				RewindTo:     2,
 			},
+		},
+	}
+
+	for _, test := range tests {
+		db := database.NewMemoryDBManager()
+		config, hash, err := test.fn(db, test.newHFBlock)
+
+		// Check the return values
+		assert.Equal(t, err, test.wantErr, test.name+": err is mismatching")
+		assert.Equal(t, config, test.wantConfig, test.name+": config is mismatching")
+		assert.Equal(t, hash, test.wantHash, test.name+": hash is mismatching")
+
+		// Check stored genesis block
+		stored := db.ReadBlock(test.wantHash, 0)
+		assert.Equal(t, stored.Hash(), test.wantHash, test.name+": stored genesis block is not compatible")
+
+		if test.wantErr == nil {
+			// Check stored chainConfig
+			storedChainConfig := db.ReadChainConfig(test.wantHash)
```

Comment:
You may want to pull up this line out of `if` statement.

Reply:
Thanks.. It is not necessary logic at all.. `if` statement is removed at https://github.com/klaytn/klaytn/pull/1017/commits/93f41f7a96224bbb924a3420e3579ab1c040f50c.